### PR TITLE
Upgrade proposal infra

### PR DESCRIPTION
**Pull request by ldjebran**
This is a proposal discussed in https://github.com/SatelliteQE/robottelo/pull/6479#discussion_r240260312
The main idea is to make the infra internal as more transparent as possible and allow to focus on the pure tests writing.

for tests demo look at test_upgrade.py and the ported test_syncplan upgrade test
for test_syncplan test output (of course the tests was run on the same sat server).
pre_upgarde
```
(robottelo) dlezz@dlezz:~/projects/robottelo-fork$ pytest -v -m 'pre_upgrade' tests/foreman/upgrade/test_syncplan.py
============================================ test session starts =============================================
platform linux -- Python 3.6.6, pytest-4.0.1, py-1.5.4, pluggy-0.8.0 -- /home/dlezz/.pyenv/versions/3.6.6/envs/robottelo/bin/python
cachedir: .pytest_cache
shared_function enabled - ON - scope: ffff - storage: redis
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: wait-for-1.0.9, xdist-1.22.5, services-1.3.1, mock-1.10.0, forked-0.2, cov-2.5.1
collecting ... 2018-12-13 16:31:50 - conftest - DEBUG - BZ deselect is disabled in settings

collected 2 items / 1 deselected                                                                             

tests/foreman/upgrade/test_syncplan.py::test_pre_sync_plan_migration PASSED                            [100%]

================================== 1 passed, 1 deselected in 36.47 seconds ===================================
```
post upgrade
```
(robottelo) dlezz@dlezz:~/projects/robottelo-fork$ pytest -v -m 'post_upgrade' tests/foreman/upgrade/test_syncplan.py
============================================ test session starts =============================================
platform linux -- Python 3.6.6, pytest-4.0.1, py-1.5.4, pluggy-0.8.0 -- /home/dlezz/.pyenv/versions/3.6.6/envs/robottelo/bin/python
cachedir: .pytest_cache
shared_function enabled - ON - scope: ffff - storage: redis
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: wait-for-1.0.9, xdist-1.22.5, services-1.3.1, mock-1.10.0, forked-0.2, cov-2.5.1
collecting ... 2018-12-13 16:35:13 - conftest - DEBUG - BZ deselect is disabled in settings

collected 2 items / 1 deselected                                                                             

tests/foreman/upgrade/test_syncplan.py::test_post_sync_plan_migration PASSED                           [100%]

=================================== 1 passed, 1 deselected in 3.48 seconds ===================================
```

For test_upgrade we are immulating scenarios where you can have a post_upgrade test that depend from many pre_upgrade tests one of them can fail and to have the dependent post upgrade dependent test skipped some pre_upgrade tests pass data to dependent post_upgrade tests . 
test_upgrade : pre_upgrade phase
```
(robottelo) dlezz@dlezz:~/projects/robottelo-fork$ pytest -v -m 'pre_upgrade' tests/foreman/upgrade/test_upgrade.py
============================================ test session starts =============================================
platform linux -- Python 3.6.6, pytest-4.0.1, py-1.5.4, pluggy-0.8.0 -- /home/dlezz/.pyenv/versions/3.6.6/envs/robottelo/bin/python
cachedir: .pytest_cache
shared_function enabled - ON - scope: ffff - storage: redis
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: wait-for-1.0.9, xdist-1.22.5, services-1.3.1, mock-1.10.0, forked-0.2, cov-2.5.1
collecting ... 2018-12-13 16:38:58 - conftest - DEBUG - BZ deselect is disabled in settings

collected 11 items / 6 deselected                                                                            

tests/foreman/upgrade/test_upgrade.py::test_capsule_pre_upgrade_1 PASSED                               [ 20%]
tests/foreman/upgrade/test_upgrade.py::test_capsule_pre_upgrade_2 FAILED                               [ 40%]
tests/foreman/upgrade/test_upgrade.py::test_capsule_pre_upgrade_2_2 ERROR                              [ 60%]
tests/foreman/upgrade/test_upgrade.py::test_capsule_pre_upgrade_3 PASSED                               [ 80%]
tests/foreman/upgrade/test_upgrade.py::test_capsule_pre_upgrade_4 PASSED                               [100%]

=================================================== ERRORS ===================================================
_______________________________ ERROR at setup of test_capsule_pre_upgrade_2_2 _______________________________

    @fixture
    def f_failing():
        # a fixture that fail
>       return 1/0
E       ZeroDivisionError: division by zero

tests/foreman/upgrade/test_upgrade.py:12: ZeroDivisionError
------------------------------------------- Captured stdout setup --------------------------------------------
2018-12-13 16:38:58 - robottelo - DEBUG - Started Test: test_upgrade.py/test_capsule_pre_upgrade_2_2
--------------------------------------------- Captured log setup ---------------------------------------------
conftest.py                135 DEBUG    Started Test: test_upgrade.py/test_capsule_pre_upgrade_2_2
================================================== FAILURES ==================================================
_________________________________________ test_capsule_pre_upgrade_2 _________________________________________

    @pre_upgrade
    def test_capsule_pre_upgrade_2():
        """A pre upgrade test that should fail"""
>       assert 1 == 0
E       assert 1 == 0

tests/foreman/upgrade/test_upgrade.py:26: AssertionError
------------------------------------------- Captured stdout setup --------------------------------------------
2018-12-13 16:38:58 - robottelo - DEBUG - Started Test: test_upgrade.py/test_capsule_pre_upgrade_2
--------------------------------------------- Captured log setup ---------------------------------------------
conftest.py                135 DEBUG    Started Test: test_upgrade.py/test_capsule_pre_upgrade_2
------------------------------------------ Captured stdout teardown ------------------------------------------
2018-12-13 16:38:58 - robottelo - DEBUG - Finished Test: test_upgrade.py/test_capsule_pre_upgrade_2
------------------------------------------- Captured log teardown --------------------------------------------
conftest.py                137 DEBUG    Finished Test: test_upgrade.py/test_capsule_pre_upgrade_2
========================= 1 failed, 3 passed, 6 deselected, 1 error in 0.04 seconds ==========================
2018-12-13 16:38:58 - robottelo - INFO - Save failed tests to file /var/tmp/robottelo_pre_upgrade_failed_tests.json
```
we have 2 tests that fail one test failure in the test it self and one failure because of setup errors 

test_upgrade: post_upgrade phase
```
(robottelo) dlezz@dlezz:~/projects/robottelo-fork$ pytest -v -m 'post_upgrade' tests/foreman/upgrade/test_upgrade.py
============================================ test session starts =============================================
platform linux -- Python 3.6.6, pytest-4.0.1, py-1.5.4, pluggy-0.8.0 -- /home/dlezz/.pyenv/versions/3.6.6/envs/robottelo/bin/python
cachedir: .pytest_cache
shared_function enabled - ON - scope: ffff - storage: redis
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: wait-for-1.0.9, xdist-1.22.5, services-1.3.1, mock-1.10.0, forked-0.2, cov-2.5.1
collecting ... 2018-12-13 16:39:02 - conftest - INFO - Deselected (because of dependant test failed): tests.foreman.upgrade.test_upgrade.test_capsule_post_upgrade_1

2018-12-13 16:39:02 - conftest - INFO - Deselected (because of dependant test failed): tests.foreman.upgrade.test_upgrade.test_capsule_post_upgrade_1_1

2018-12-13 16:39:02 - conftest - INFO - Deselected (because of dependant test failed): tests.foreman.upgrade.test_upgrade.test_capsule_post_upgrade_1_2

2018-12-13 16:39:02 - conftest - DEBUG - BZ deselect is disabled in settings

collected 11 items / 8 deselected                                                                            

tests/foreman/upgrade/test_upgrade.py::test_capsule_post_upgrade_2 PASSED                              [ 33%]
tests/foreman/upgrade/test_upgrade.py::test_capsule_post_upgrade_3 PASSED                              [ 66%]
tests/foreman/upgrade/test_upgrade.py::test_capsule_post_upgrade_4 PASSED                              [100%]

=================================== 3 passed, 8 deselected in 0.02 seconds ===================================
```
In post upgrade we assert that the pre_upgrade tests data was passed successfully to post upgrade tests, we can notice that the dependent tests from pre_upgarde tests phase that failed was skipped and the tests with no dependency passed also.

Note upgrade conftest is xdist compatibe, so the tests can run on many threads:

tests output with xdist 4 workers

pre_upgrade test phase:
```
(robottelo) dlezz@dlezz:~/projects/robottelo-fork$ pytest -v -n 4 -m 'pre_upgrade' tests/foreman/upgrade/
============================================ test session starts =============================================
platform linux -- Python 3.6.6, pytest-4.0.1, py-1.5.4, pluggy-0.8.0 -- /home/dlezz/.pyenv/versions/3.6.6/envs/robottelo/bin/python
cachedir: .pytest_cache
shared_function enabled - ON - scope: ffff - storage: redis
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: wait-for-1.0.9, xdist-1.22.5, services-1.3.1, mock-1.10.0, forked-0.2, cov-2.5.1
[gw0] linux Python 3.6.6 cwd: /home/dlezz/projects/robottelo-fork
[gw1] linux Python 3.6.6 cwd: /home/dlezz/projects/robottelo-fork
[gw2] linux Python 3.6.6 cwd: /home/dlezz/projects/robottelo-fork
[gw3] linux Python 3.6.6 cwd: /home/dlezz/projects/robottelo-fork
[gw0] Python 3.6.6 (default, Jul 31 2018, 17:47:38)  -- [GCC 5.4.0 20160609]
[gw1] Python 3.6.6 (default, Jul 31 2018, 17:47:38)  -- [GCC 5.4.0 20160609]
[gw2] Python 3.6.6 (default, Jul 31 2018, 17:47:38)  -- [GCC 5.4.0 20160609]
[gw3] Python 3.6.6 (default, Jul 31 2018, 17:47:38)  -- [GCC 5.4.0 20160609]
gw0 [6] / gw1 [6] / gw2 [6] / gw3 [6]
scheduling tests via LoadScheduling

tests/foreman/upgrade/test_syncplan.py::test_pre_sync_plan_migration 
tests/foreman/upgrade/test_upgrade.py::test_capsule_pre_upgrade_2_2 
tests/foreman/upgrade/test_upgrade.py::test_capsule_pre_upgrade_1 
tests/foreman/upgrade/test_upgrade.py::test_capsule_pre_upgrade_2 
[gw1] [ 16%] PASSED tests/foreman/upgrade/test_upgrade.py::test_capsule_pre_upgrade_1 
tests/foreman/upgrade/test_upgrade.py::test_capsule_pre_upgrade_4 
[gw1] [ 33%] PASSED tests/foreman/upgrade/test_upgrade.py::test_capsule_pre_upgrade_4 
[gw3] [ 50%] ERROR tests/foreman/upgrade/test_upgrade.py::test_capsule_pre_upgrade_2_2 
[gw2] [ 66%] FAILED tests/foreman/upgrade/test_upgrade.py::test_capsule_pre_upgrade_2 
[gw0] [ 83%] PASSED tests/foreman/upgrade/test_syncplan.py::test_pre_sync_plan_migration 
tests/foreman/upgrade/test_upgrade.py::test_capsule_pre_upgrade_3 
[gw0] [100%] PASSED tests/foreman/upgrade/test_upgrade.py::test_capsule_pre_upgrade_3 

=================================================== ERRORS ===================================================
_______________________________ ERROR at setup of test_capsule_pre_upgrade_2_2 _______________________________
[gw3] linux -- Python 3.6.6 /home/dlezz/.pyenv/versions/3.6.6/envs/robottelo/bin/python

    @fixture
    def f_failing():
        # a fixture that fail
>       return 1/0
E       ZeroDivisionError: division by zero

tests/foreman/upgrade/test_upgrade.py:12: ZeroDivisionError
------------------------------------------- Captured stdout setup --------------------------------------------
2018-12-13 17:13:12 - robottelo - INFO - Entities cleaner disabled
2018-12-13 17:13:12 - robottelo - DEBUG - Started Test: test_upgrade.py/test_capsule_pre_upgrade_2_2
--------------------------------------------- Captured log setup ---------------------------------------------
conftest.py                126 INFO     Entities cleaner disabled
conftest.py                135 DEBUG    Started Test: test_upgrade.py/test_capsule_pre_upgrade_2_2
================================================== FAILURES ==================================================
_________________________________________ test_capsule_pre_upgrade_2 _________________________________________
[gw2] linux -- Python 3.6.6 /home/dlezz/.pyenv/versions/3.6.6/envs/robottelo/bin/python

    @pre_upgrade
    def test_capsule_pre_upgrade_2():
        """A pre upgrade test that should fail"""
>       assert 1 == 0
E       assert 1 == 0

tests/foreman/upgrade/test_upgrade.py:26: AssertionError
------------------------------------------- Captured stdout setup --------------------------------------------
2018-12-13 17:13:12 - robottelo - INFO - Entities cleaner disabled
2018-12-13 17:13:12 - robottelo - DEBUG - Started Test: test_upgrade.py/test_capsule_pre_upgrade_2
--------------------------------------------- Captured log setup ---------------------------------------------
conftest.py                126 INFO     Entities cleaner disabled
conftest.py                135 DEBUG    Started Test: test_upgrade.py/test_capsule_pre_upgrade_2
------------------------------------------ Captured stdout teardown ------------------------------------------
2018-12-13 17:13:12 - robottelo - DEBUG - Finished Test: test_upgrade.py/test_capsule_pre_upgrade_2
------------------------------------------- Captured log teardown --------------------------------------------
conftest.py                137 DEBUG    Finished Test: test_upgrade.py/test_capsule_pre_upgrade_2
================================ 1 failed, 4 passed, 1 error in 34.88 seconds ================================
2018-12-13 17:13:10 - conftest - DEBUG - Registering custom pytest_namespace

2018-12-13 17:13:45 - robottelo - INFO - Save failed tests to file /var/tmp/robottelo_pre_upgrade_failed_tests.json

```

post_upgrade phase:
```
(robottelo) dlezz@dlezz:~/projects/robottelo-fork$ pytest -v -n 4 -m 'post_upgrade' tests/foreman/upgrade/
============================================ test session starts =============================================
platform linux -- Python 3.6.6, pytest-4.0.1, py-1.5.4, pluggy-0.8.0 -- /home/dlezz/.pyenv/versions/3.6.6/envs/robottelo/bin/python
cachedir: .pytest_cache
shared_function enabled - ON - scope: ffff - storage: redis
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: wait-for-1.0.9, xdist-1.22.5, services-1.3.1, mock-1.10.0, forked-0.2, cov-2.5.1
[gw0] linux Python 3.6.6 cwd: /home/dlezz/projects/robottelo-fork
[gw1] linux Python 3.6.6 cwd: /home/dlezz/projects/robottelo-fork
[gw2] linux Python 3.6.6 cwd: /home/dlezz/projects/robottelo-fork
[gw3] linux Python 3.6.6 cwd: /home/dlezz/projects/robottelo-fork
[gw0] Python 3.6.6 (default, Jul 31 2018, 17:47:38)  -- [GCC 5.4.0 20160609]
[gw1] Python 3.6.6 (default, Jul 31 2018, 17:47:38)  -- [GCC 5.4.0 20160609]
[gw2] Python 3.6.6 (default, Jul 31 2018, 17:47:38)  -- [GCC 5.4.0 20160609]
[gw3] Python 3.6.6 (default, Jul 31 2018, 17:47:38)  -- [GCC 5.4.0 20160609]
gw0 [4] / gw1 [4] / gw2 [4] / gw3 [4]
scheduling tests via LoadScheduling

tests/foreman/upgrade/test_upgrade.py::test_capsule_post_upgrade_2 
tests/foreman/upgrade/test_syncplan.py::test_post_sync_plan_migration 
tests/foreman/upgrade/test_upgrade.py::test_capsule_post_upgrade_3 
tests/foreman/upgrade/test_upgrade.py::test_capsule_post_upgrade_4 
[gw1] [ 25%] PASSED tests/foreman/upgrade/test_upgrade.py::test_capsule_post_upgrade_2 
[gw3] [ 50%] PASSED tests/foreman/upgrade/test_upgrade.py::test_capsule_post_upgrade_4 
[gw2] [ 75%] PASSED tests/foreman/upgrade/test_upgrade.py::test_capsule_post_upgrade_3 
[gw0] [100%] PASSED tests/foreman/upgrade/test_syncplan.py::test_post_sync_plan_migration 

========================================== 4 passed in 4.98 seconds ==========================================
2018-12-13 17:14:27 - conftest - DEBUG - Registering custom pytest_namespace

(robottelo) dlezz@dlezz:~/projects/robottelo-fork$ 
```
Note the info about skipped is not shown on the console (with xdist) but shown in the log file.

### CODE REVIEW

File: tests/foreman/upgrade/conftest.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+        date=now.strftime("%Y-%m-%d %H:%M:%S"),
+        level=level,
+        message=message
+    )
+    print(full_message)  # noqa
+    with open('robottelo.log', 'a') as log_file:
+        log_file.write(full_message)
+
+
+# todo remove when upgrade_tests will be python 3 compatible
```
Comment (jyejare) [inline tests/foreman/upgrade/conftest.py]: @ldjebran , PRs are now merged to run upgrade scenarios on python3.

**Comment by jyejare** [comment]
@ldjebran , The changes look good to me. If you can populate docstrings on missing functions, it would help folks to review and use.

**Comment by ldjebran** [comment]
@jyejare Finalized , added support for sipped tests in pre_upgrade tests , so if a test is skipped in pre_upgrade it's dependent post_upgrade tests are also skipped.

pre upgrade
```
(robottelo) dlezz@dlezz:~/projects/robottelo-fork$ pytest -v -m "pre_upgrade" tests/foreman/upgrade/test_syncplan.py
============================================ test session starts =============================================
platform linux -- Python 3.6.6, pytest-4.0.2, py-1.5.4, pluggy-0.8.0 -- /home/dlezz/.pyenv/versions/robottelo/bin/python3.6
cachedir: .pytest_cache
shared_function enabled - ON - scope: ffff - storage: redis
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: wait-for-1.0.9, xdist-1.22.5, timeout-1.3.3, services-1.3.1, mock-1.10.0, forked-0.2, cov-2.5.1
collecting ... 2019-01-22 14:53:46 - conftest - DEBUG - BZ deselect is disabled in settings

collected 2 items / 1 deselected                                                                             

tests/foreman/upgrade/test_syncplan.py::test_pre_sync_plan_migration PASSED                            [100%]

================================== 1 passed, 1 deselected in 40.62 seconds ===================================
2019-01-22 14:54:26 - robottelo - INFO - Save failed tests to file /var/tmp/robottelo_pre_upgrade_failed_tests.json
```

post_upgrade
```
(robottelo) dlezz@dlezz:~/projects/robottelo-fork$ pytest -v -m "post_upgrade" tests/foreman/upgrade/test_syncplan.py
============================================ test session starts =============================================
platform linux -- Python 3.6.6, pytest-4.0.2, py-1.5.4, pluggy-0.8.0 -- /home/dlezz/.pyenv/versions/robottelo/bin/python3.6
cachedir: .pytest_cache
shared_function enabled - ON - scope: ffff - storage: redis
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: wait-for-1.0.9, xdist-1.22.5, timeout-1.3.3, services-1.3.1, mock-1.10.0, forked-0.2, cov-2.5.1
collecting ... 2019-01-22 14:54:37 - conftest - DEBUG - BZ deselect is disabled in settings

collected 2 items / 1 deselected                                                                             

tests/foreman/upgrade/test_syncplan.py::test_post_sync_plan_migration PASSED                           [100%]

=================================== 1 passed, 1 deselected in 9.00 seconds ===================================
```

File: tests/foreman/upgrade/conftest.py (hunk 1 above)
Comment (ldjebran) [reply tests/foreman/upgrade/conftest.py]: @jyejare still not compatible as having error when importing
```
  from upgrade_tests.helpers.scenarios import create_dict, get_entity_data
E     File "/home/dlezz/.pyenv/versions/robottelo/lib/python3.6/site-packages/upgrade_tests/helpers/scenarios.py", line 164
E       print hammer.hammer('subscription upload --file {0} '


```

Comment (jyejare) [reply tests/foreman/upgrade/conftest.py]: Yeah @ldjebran , That is the scenario yet to be moved to py3. I think @vijay8451 is going to have a look into it else I will.

**Comment by jyejare** [comment]
I think if everyone is ok with this implementation, lets get this in and use it.

**Comment by san7ket** [comment]
@omkarkhatavkar @jyejare I think we can merge it

**Comment by jyejare** [comment]
@omkarkhatavkar @san7ket , Issue https://github.com/SatelliteQE/robottelo/issues/6847 is raised to update the scenarios with this implementation.

Merging this !

**Pull request merged**